File: macos/Tests/ElevateCoreTests/QuickActivateTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct QuickActivateTests {
    let key = RoleKey(identityId: "i", tenantId: "t", scope: .entraDirectory(roleDefinitionId: "r", directoryScopeId: "/"))
    func role(justification: Bool = true, ticket: Bool = false, approval: Bool = false, mfa: Bool = true, max: Int = 4) -> EligibleRole {
        EligibleRole(key: key, displayName: "R", source: .discovered,
                     policy: RolePolicy(defaultDuration: .seconds(3600), maximumDuration: .seconds(max * 3600), requiresJustification: justification,
                                        requiresTicket: ticket, requiresMFA: mfa, requiresApproval: approval, authenticationContext: "c1"))
    }
    let memory = RoleMemory(roleKey: RoleKey(identityId: "i", tenantId: "t", scope: .entraDirectory(roleDefinitionId: "r", directoryScopeId: "/")), justification: "INC-1", lastDuration: .seconds(8 * 3600))

    @Test func readyUsesRememberedReasonAndCappedDuration() {
        guard case .ready(let reqs) = QuickActivate.decide(role: role(), memory: memory) else { Issue.record("expected ready"); return }
        #expect(reqs.count == 1 && reqs[0].justification == "INC-1" && reqs[0].duration == .seconds(4 * 3600))
        #expect(reqs[0].authenticationContext == "c1" && reqs[0].startDateTime == nil && reqs[0].ticket == nil)
    }

    @Test func dialogReasons() {
        #expect(QuickActivate.decide(role: role(), memory: nil) == .needsDialog("no remembered reason"))
        #expect(QuickActivate.decide(role: role(justification: false), memory: nil) != .needsDialog("no remembered reason"))
        #expect(QuickActivate.decide(role: role(ticket: true), memory: memory) == .needsDialog("ticket required"))
        #expect(QuickActivate.decide(role: role(approval: true), memory: memory) == .needsDialog("approval required"))
    }

    @Test func profileDecision() {
        let r = role()
        let ok = ProfilePlanItem(roleKey: key, role: r, duration: .seconds(3600), disposition: .activate)
        let skipped = ProfilePlanItem(roleKey: key, role: r, duration: .seconds(3600), disposition: .alreadyActive)
        guard case .ready(let reqs) = QuickActivate.decide(items: [ok, skipped], justification: "INC-2") else { Issue.record("expected ready"); return }
        #expect(reqs.count == 1 && reqs[0].justification == "INC-2")
        #expect(QuickActivate.decide(items: [ok], justification: nil) == .needsDialog("no remembered reason"))
        #expect(QuickActivate.decide(items: [ProfilePlanItem(roleKey: key, role: nil, duration: .seconds(60), disposition: .notLoaded)], justification: "x") == .needsDialog("roles still loading"))
        #expect(QuickActivate.decide(items: [ProfilePlanItem(roleKey: key, role: role(ticket: true), duration: .seconds(60), disposition: .activate)], justification: "x") == .needsDialog("ticket required"))
        #expect(QuickActivate.decide(items: [skipped], justification: "x") == .needsDialog("nothing to activate"))
    }
}

File: windows/tests/Elevate.Core.Tests/QuickActivateTests.cs
using Elevate.Core.Coordination;
using Elevate.Core.Models;
using Elevate.Core.Storage;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>QuickActivateTests</c>.</summary>
public class QuickActivateTests
{
    private static readonly RoleKey Key = new("i", "t", new EntraDirectoryScope("r", "/"));
    private static readonly RoleMemory Memory = new(Key, "INC-1", TimeSpan.FromHours(8));

    private static EligibleRole Role(bool justification = true, bool ticket = false, bool approval = false, bool mfa = true, int max = 4) =>
        new(Key, "R", RoleSource.Discovered, new RolePolicy(
            TimeSpan.FromHours(1), TimeSpan.FromHours(max), justification, ticket, mfa, approval, "c1"));

    [Fact]
    public void ReadyUsesRememberedReasonAndCappedDuration()
    {
        var decision = QuickActivate.Decide(Role(), Memory);

        var ready = decision.Should().BeOfType<QuickActivateDecision.Ready>().Subject;
        ready.Requests.Should().HaveCount(1);
        var request = ready.Requests[0];
        request.Justification.Should().Be("INC-1");
        request.Duration.Should().Be(TimeSpan.FromHours(4));
        request.AuthenticationContext.Should().Be("c1");
        request.StartDateTime.Should().BeNull();
        request.Ticket.Should().BeNull();
    }

    [Fact]
    public void DialogReasons()
    {
        QuickActivate.Decide(Role(), null).Should().Be(new QuickActivateDecision.NeedsDialog("no remembered reason"));
        QuickActivate.Decide(Role(justification: false), null).Should().NotBe(new QuickActivateDecision.NeedsDialog("no remembered reason"));
        QuickActivate.Decide(Role(ticket: true), Memory).Should().Be(new QuickActivateDecision.NeedsDialog("ticket required"));
        QuickActivate.Decide(Role(approval: true), Memory).Should().Be(new QuickActivateDecision.NeedsDialog("approval required"));
    }

    [Fact]
    public void ProfileDecision()
    {
        var r = Role();
        var ok = new ProfilePlanItem(Key, r, TimeSpan.FromHours(1), ProfilePlanDisposition.Activate);
        var skipped = new ProfilePlanItem(Key, r, TimeSpan.FromHours(1), ProfilePlanDisposition.AlreadyActive);

        var ready = QuickActivate.Decide([ok, skipped], "INC-2").Should().BeOfType<QuickActivateDecision.Ready>().Subject;
        ready.Requests.Should().HaveCount(1);
        ready.Requests[0].Justification.Should().Be("INC-2");

        QuickActivate.Decide([ok], null).Should().Be(new QuickActivateDecision.NeedsDialog("no remembered reason"));
        QuickActivate.Decide([new ProfilePlanItem(Key, null, TimeSpan.FromMinutes(1), ProfilePlanDisposition.NotLoaded)], "x")
            .Should().Be(new QuickActivateDecision.NeedsDialog("roles still loading"));
        QuickActivate.Decide([new ProfilePlanItem(Key, Role(ticket: true), TimeSpan.FromMinutes(1), ProfilePlanDisposition.Activate)], "x")
            .Should().Be(new QuickActivateDecision.NeedsDialog("ticket required"));
        QuickActivate.Decide([skipped], "x").Should().Be(new QuickActivateDecision.NeedsDialog("nothing to activate"));
    }
}
